openraft/src/testing/store_builder.rs

```diff
@@ -18,24 +19,10 @@ where
     C: RaftTypeConfig,
     S: RaftStorage<C>,
 {
-    async fn build(&self) -> S;
-}
-
-/// Make the tests easy to use by converting a closure to a [`StoreBuilder`].
-///
-/// E.g. to run tests on your [`RaftStorage`] implementation, just use `Suite::test_all(|| new_store())`,
-/// if your have already provided `async fn new_store() -> MyStore`
-#[async_trait]
-impl<C, S, Fu, F> StoreBuilder<C, S> for F
-where
-    C: RaftTypeConfig,
-    S: RaftStorage<C>,
-    Fu: Future<Output = S> + Send,
-    F: Fn() -> Fu + Sync + Send,
-{
-    async fn build(&self) -> S {
-        (self)().await
-    }
+    async fn run_test<Fun, Ret, Res>(&self, t: Fun) -> Result<Ret, StorageError<C::NodeId>>
+    where
+        Res: Future<Output = Result<Ret, StorageError<C::NodeId>>> + Send,
+        Fun: Fn(S) -> Res + Sync + Send;
```

Comment:
Ja, that was my first thought too let me explain why I decided against it and went with changing it to move to a call-back style instead.

The original challenge I had was that the suite did make the assumption that any store is stateless and can be re-created many times without issue. Returning the dir was my first thought and I experimented with it the three reasons I decided against it are:

1) Returning a context w/ store + temp dir, would change the assumption to "stores are stateless or their only state is in a temp dir". So I worried that when tests get more complex, or we test stores that don't fall into this category would then again need a re-write of the logic. With the callback logic, the builder has full control over setup and cleanup, so the framework makes no assumptions so this becomes quite flexible. So to sum up: I felt the callback solution is more flexible and future prove for other use cases / more complex tests

2) returning the context involves the same amount of changes to all the tests, as each would need to adapt to having a context instead of the store returned. To sum this up: The difference in change isn't that big.

3) Rust scopes, they are tricky, and relying on a behavior that isn't widely understood or able to be visualized / explained seems wrong to me. The example works because it's a simple function but based on the specification, as written, [rust has the right to drop the temp directory at any point in the function body](https://doc.rust-lang.org/reference/destructors.html) - when it does so is neither predictable nor controllable by a user unless they keep a firm handle on the state and ensure dropping in the right place via `drop()` - this would make the change more verbose and significantly more error-prone to write tests. To sum this up: The callback model is a lot less complex to reason about and less error-prone.

I hope that makes sense?

Reply:
Make sense. I'm convinced:D